[pigeon] Adds Swift support for Pigeon

ailtonvivaz opened the pull request:
This PR adds Swift support for Pigeon. This code was based on the Java generator.

Fixes https://github.com/flutter/flutter/issues/66499

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

stuartmorgan commented on packages/pigeon/lib/swift_generator.dart:
```diff
@@ -0,0 +1,642 @@
+                  'let error = FlutterError(code: "unexpected-args-count", message: "Unexpected argument parameters count", details: "Expected parameters count: ${method.arguments.length}. Received: \\(args.count)")');
+              indent.writeln('reply(wrapError(error))');
+              indent.writeln('return');
+            });
```
These to checks contribute to a ton of boilerplate in the generated code for a long list of methods. Per some recent discussion in `video_player` reviews, we should use helpers when possible. These two checks could be moved to a helper (taking the expected count as a parameter), cutting down on the repeated generated code.

stuartmorgan replied:
Although actually, why is the length a runtime check at all? It's not clear to me that we want to add a runtime cost of checks that can only fail if someone corrupts their generated code (e.g., by only updating one side).

ailtonvivaz replied:
I got it. There is no reason to keep this runtime check. I remove it.

But this make me thinking about another check: the null safe check.
Is there need to this? Or can we presume that the type is correct either?

Follow some proposes to change in generator.
The force unwrap (`as!`) would be used just in case the type be non-null (`Type`).
To null type (`Type?`), would be used `as?`.

## Arguments parser
This
```swift
let valueArg = args[0] as! Type
```
instead of
```swift
guard let valueArg = args[0] as? Type else {
  let error = FlutterError(code: "unexpected-arg-type", message: "value argument unexpected type", details: "Expected type: Type. Received: \(type(of: args[0]))")
  reply(wrapError(error))
  return
}
```

## .fromMap
This
```swift
let value = map["value"] as! Type
```
instead of
```swift
guard let value = map["value"] as? Type else { return nil }
```

## CodeReader: FlutterStandardReader
This
```swift
return Type.fromMap(self.readValue() as! [String: Any?])  
```
instead of
```swift
guard let map = self.readValue() as? [String: Any?] else {
  return nil
}
return Type.fromMap(map)
```

stuartmorgan commented on packages/pigeon/lib/swift_generator.dart:
```diff
@@ -0,0 +1,642 @@
+              indent.write('guard let $argName = args[$index] as? $argType ');
+              indent.scoped('else {', '}', () {
+                indent.writeln(
+                    'let error = FlutterError(code: "unexpected-arg-type", message: "${arg.name} argument unexpected type", details: "Expected type: $argType. Received: \\(type(of: args[$index]))")');
```
Similarly, the creation of this error could be a small helper that takes these three variables as parameters, so the string isn't repeated over and over.

And in fact, that will reduce actual binary size, because otherwise every one of these entire strings is duplicated within the binary since they are all slightly different.

ailtonvivaz replied:
I got it. I'll review the generator to refactor and avoid those boilerplates.
But, in this case I think similarly above. Instead use a helper, what if remove this?
Is there need to this? Or can we presume that the type is correct either?

Follow some proposes to change in generator.
The force unwrap (`as!`) would be used just in case the type be non-null (`Type`).
To null type (`Type?`), would be used `as?`.

## Arguments parser
This
```swift
let valueArg = args[0] as! Type
```
instead of
```swift
guard let valueArg = args[0] as? Type else {
  let error = FlutterError(code: "unexpected-arg-type", message: "value argument unexpected type", details: "Expected type: Type. Received: \(type(of: args[0]))")
  reply(wrapError(error))
  return
}
```

## .fromMap
This
```swift
let value = map["value"] as! Type
```
instead of
```swift
guard let value = map["value"] as? Type else { return nil }
```

## CodeReader: FlutterStandardReader
This
```swift
return Type.fromMap(self.readValue() as! [String: Any?])  
```
instead of
```swift
guard let map = self.readValue() as? [String: Any?] else {
  return nil
}
return Type.fromMap(map)
```

stuartmorgan replied:
I believe we force-cast on the Dart side, so that's probably reasonable here as well. I'll defer to @gaaclarke on that.

ailtonvivaz replied:
I changed it

stuartmorgan reviewed (changes requested):
Thanks for the contribution!

@gaaclarke should do the main review, but I did a quick high-level pass over the generator code itself.

ailtonvivaz commented on packages/pigeon/lib/swift_generator.dart:
```diff
@@ -0,0 +1,642 @@
+  });
+
+  /// Prefix that will be appended before all generated classes and protocols.
+  final String? prefix;
```
In fact, Swift doesn't.
The only reason for me add that was for testing. Unlike other languages (ObjC, Java, Kotlin) Swift does't include/import classes by file.
And, since the sample pigeons has classes with same names, happens a conflict if there not be a prefix.
Unfortunately, this was the only workaround I found for this at the moment.

stuartmorgan replied:
I don't follow; what is the collision here that can't be resolved with module names (which is how collisions are supposed to be resolved in Swift, thus the lack of prefixing)?

ailtonvivaz replied:
I agree.
But in this case, all classes are in the same module.
The script generates the files `platform_tests/ios_swift_unit_tests/ios/Runner/*.gen.swift` to test.
But, as `pigeons/flutter_unittests.dart` as `pigeons/host2flutter.dart` has the class `SearchReply` the Swift compile fail with `Swift Compiler Error (Xcode): Invalid redeclaration of 'SearchReply'`
There are other classes with this conflict.

A workaround for this would be rename this pigeons classes in order to make them different among theirs.
Can I proceed with that?

stuartmorgan replied:
If the problem is the unit tests, we should absolutely change the unit tests to work rather than add functionality to the public API that should not exist.

ailtonvivaz replied:
Done. I've removed the prefix feature for Swift generator and change the names of pigeons data classes and apis.

ailtonvivaz commented on packages/pigeon/lib/swift_generator.dart:
```diff
@@ -0,0 +1,642 @@
+  final String apiName = _className(prefix, api.name);
+
+  indent.writeln(
+      '/** Generated protocol from Pigeon that represents a handler of messages from Flutter.*/');
```
Thanks. I fixed that.

ailtonvivaz commented on packages/pigeon/lib/swift_generator.dart:
```diff
@@ -0,0 +1,642 @@
+/// Example:
+/// protocol Foo {
+///   Int add(x: Int, y: Int);
+///   static func setup(FlutterBinaryMessenger binaryMessenger, Foo api) {...}
```
Thanks. I change it.

ailtonvivaz commented on packages/pigeon/lib/swift_generator.dart:
```diff
@@ -0,0 +1,642 @@
+        '/** Sets up an instance of `$apiName` to handle messages through the `binaryMessenger`. */');
+    indent.write(
+        'static func setup(binaryMessenger: FlutterBinaryMessenger, api: $apiName) ');
+    // aceitar api null
```
My mistake keeps this reminder in the code. I remove it and I've implemented this: Allow nil api to remove channel handler.

ailtonvivaz commented on packages/pigeon/lib/swift_generator.dart:
```diff
@@ -0,0 +1,642 @@
+      final String returnType = func.returnType.isVoid
+          ? ''
+          : _swiftTypeForDartType(prefix, func.returnType);
+      final String nullsafe = func.returnType.isNullable ? '?' : '';
```
Thanks, I fixed it.

ailtonvivaz commented on packages/pigeon/lib/swift_generator.dart:
```diff
@@ -0,0 +1,642 @@
+        indent.write('');
+        for (final EnumeratedClass customClass in getCodecClasses(api, root)) {
+          indent.write(
+              'if let value = value as? ${_className(prefix, customClass.name)} ');
```
Really. I fixed it.

ailtonvivaz replied:
I don't think so. At the moment, pigeon generator doesn't support inheritance generation.
In this case, I don't see any risk with it.

Using `type(of:)`, It need to check the type and to cast.
```swift
if type(of: value), let value = value as! Type { ... }
```

I propose keep this approach and improve the readability, like this:
```swift
switch value {
case let value as Type:
...
}
```

ailtonvivaz replied:
I completely agree with you.
I'll fix it.

ailtonvivaz replied:
I've tried this change but it isn't working.
When I use `type(of:)` it returns `__SwiftValue` instead `Type.self`.
If I compare `type(of: value) == Type.self` it returns `false`.
With this, I got the follow error: `Unsupported value for standard codec (NSInternalInconsistencyException)`

ailtonvivaz commented on packages/pigeon/lib/swift_generator.dart:
```diff
@@ -0,0 +1,642 @@
+        indent.write('');
+        for (final EnumeratedClass customClass in getCodecClasses(api, root)) {
+          indent.write(
+              'if let value = value as? ${_className(prefix, customClass.name)} ');
```
Really. I fixed it.

ailtonvivaz commented on packages/pigeon/lib/swift_generator.dart:
```diff
@@ -0,0 +1,642 @@
+      '/** Generated protocol from Pigeon that represents a handler of messages from Flutter.*/');
+  indent.write('protocol $apiName ');
+  indent.scoped('{', '}', () {
+    for (final Method method in api.methods) {
```
Really. Nice idea.
I think in an annotation in each parameter to define this.

 How about that?
```dart
@HostApi()
abstract class Api {
  String? getString(@ParameterLabel('for') String key);
  void setString(@ParameterLabel('_') String value, @ParameterLabel('for') String key);
  void remove(String key);
}
```

This would generate

### ObjC
```objc
@protocol Api
- (nullable NSString *)getStringForKey:(NSString *)key error:(FlutterError *_Nullable *_Nonnull)error;
/// @return `nil` only when `error != nil`.
- (void)setString:(NSString *)value forKey:(NSString *)key error:(FlutterError *_Nullable *_Nonnull)error;
/// @return `nil` only when `error != nil`.
- (void)removeKey:(NSString *)key error:(FlutterError *_Nullable *_Nonnull)error;
@end
```

### Swift
```swift
protocol Api {
  func getString(for key: String) -> String?
  func setString(_ value: String, for key: String)
  func remove(key: String)
}
```

This samples are about HostApi, but make more sense FlutterApi because the native code will consume.

ailtonvivaz replied:
> I would expect it to work the same way as `@ObjCSelector` so that there aren't two completely different ways of doing essentially the same thing.

I've be thinking about this. How about we change this annotation name? (of course, duplicate and deprecate)
Maybe to `MethodSelector`, `Selector`?
Thus, it could be used both to ObjC and Swift.

gaaclarke replied:
Yea, I thought about that.  I don't like those names since "selector" is objc specific.  You don't technically have one in Swift.

I don't want you to get bogged down with this decision.  Maybe for now we can just reuse `@ObjcSelector` and consider refactoring it in a different PR?  Or better yet, just don't support this yet and file an issue.  @stuartmorgan alright said it's fine to defer this for now too.  There is already a lot to review in this one PR.

stuartmorgan replied:
To be clear, I wasn't suggesting a single annotation that covers both. The naming conventions for ObjC and Swift are not always going to be the same, so they should be separately controllable. I just meant it shouldn't be param-based here, and method-based for ObjC. We should follow the pattern established by the ObjC annotation for the new Swift annotation.

ailtonvivaz replied:
Ok, I got it.
I could support this annotation and open a file a new issue, but i dont know how proceed to create a issue to a feature that even not exists.
Should I wait this PR be merged to file a issue to discuss a new annotation to Swift?

stuartmorgan replied:
> I could support this annotation and open a file a new issue, but i dont know how proceed to create a issue to a feature that even not exists.

It's fine to open an issue that says something like "PR #XYZ will add Swift support, but [...]".

ailtonvivaz replied:
Issue created: https://github.com/flutter/flutter/issues/105932

gaaclarke commented on packages/pigeon/lib/swift_generator.dart:
```diff
@@ -0,0 +1,645 @@
+/// class Foo {
+///   private let binaryMessenger: FlutterBinaryMessenger
+///   init(binaryMessenger: FlutterBinaryMessenger) {...}
+///   func add(x: 32, y: Int32, completion: @escaping (Int32?) -> Void) {...}
```
s/32/Int32

ailtonvivaz replied:
Fixed

gaaclarke commented on packages/pigeon/lib/swift_generator.dart:
```diff
@@ -0,0 +1,645 @@
+    }
+
+    void writeToMap() {
+      indent.write('func toMap() -> [String: Any?] ');
```
We have to make sure that this isn't public.  We want the ability to change the wire format of pigeon so we can't let people rely on this.  I think in swift's case we can just make these methods `internal`.

ailtonvivaz replied:
As this method as `.fromMap` are used in tests to validate the conversion, so do in ObjC.
By default, it is internal in Swift. Thus, any code outside the module couldn't access that.

gaaclarke reviewed (commented):
This is looking awesome, thanks.  I'm going to download it and play with it a bit more since it's easier if I can see the generated code.  How the code is generated looks good and there isn't any glaring omission at first glance.  The big thing will be ensuring that the platform_tests coverage is good.  Don't worry about e2e tests.  Let's focus on the platform_tests as the gold standard.

gaaclarke commented on packages/pigeon/run_tests.sh:
```diff
@@ -343,6 +390,7 @@ while getopts "t:l?h" opt; do
     flutter_unittests) should_run_flutter_unittests=true ;;
     ios_e2e_tests) should_run_ios_e2e_tests=true ;;
     ios_unittests) should_run_ios_unittests=true ;;
+    ios_swift_unittests) should_run_ios_swift_unittests=true ;;
```
Can you add this to the `-l` output as well where the tests are listed please?

ailtonvivaz replied:
Done

gaaclarke commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/EnumTests.swift:
```diff
@@ -0,0 +1,23 @@
+
+    func testEcho() throws {
+        let data = ACData(state: .Error)
+        let binaryMessender = EchoBinaryMessenger(codec: ACEnumApi2HostCodec.shared)
```
s/binaryMessender/binaryMessenger

ailtonvivaz replied:
Fixed

gaaclarke reviewed (changes requested):
Can you also add a platform_test that exercises the HostApi?  I think there is just one in the async tests.  I noticed there wasn't much coverage there and added one in the PR that is about to land: https://github.com/flutter/packages/pull/895/files?show-viewed-files=true&file-filters%5B%5D=#diff-0591c0be35756f8c6961ad4e999d94bf45b3a1aac48500a519259c264af47c3bR48

I'll try to land that today so that you have that as an example to build off of.

ailtonvivaz commented:
> This is looking awesome, thanks. I'm going to download it and play with it a bit more since it's easier if I can see the generated code. How the code is generated looks good and there isn't any glaring omission at first glance. The big thing will be ensuring that the platform_tests coverage is good. Don't worry about e2e tests. Let's focus on the platform_tests as the gold standard.

I've just ad more host test cases

ailtonvivaz commented:
> Can you also add a platform_test that exercises the HostApi? I think there is just one in the async tests. I noticed there wasn't much coverage there and added one in the PR that is about to land: https://github.com/flutter/packages/pull/895/files?show-viewed-files=true&file-filters%5B%5D=#diff-0591c0be35756f8c6961ad4e999d94bf45b3a1aac48500a519259c264af47c3bR48
> 
> I'll try to land that today so that you have that as an example to build off of.

I've merged the new feature (nullable arguments) and updated the swift generator to support this either.

gaaclarke commented:
(merged master)

gaaclarke commented:
@ailtonvivaz I'm going to play around with this PR locally this afternoon and try to get all the CI steps clean for you and add any tests I think might be missing if you don't mind.  That's what I did for the C++ PR and it gave me a better understanding of everything.

gaaclarke reviewed (approved):
I've looked over this locally and double checked that all the tests were running as we expected and everything looks good.  This generator is as complete as any other and matches the tests we have for something like objc.  The generator matches the style of the other ones so I can't fault it there either.  Thanks so much for the contribution.

The only question I have is @stuartmorgan do we want to put this behind `experimental`?  Just to make sure we didn't miss some style problem with Swift?

gaaclarke commented:
Ahh one thing I'm realizing though is this is missing some of the newer null-safety support, like non-null fields.  I think we should make the flag `--experimental_swift_out` and we should add another phase where generators can authenticate the AST so they can report lapses in coverage.  I can do that in a separate PR.

gaaclarke commented:
I added a validation phase that blocks nonnull fields on classes.  That's the only omission I saw at first glance.  I'll switch the flags to `--experimental_swift_out` too next.

gaaclarke commented:
I just did `checkout --ours` for merge conflicts, hoping that they would show up unresolved in github, but they showed up as overwriting them all with our version.  I'll fix that.

ailtonvivaz commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -511,14 +511,27 @@ pigeonMap['${field.name}'] != null
+            if (isDouble) {
+              indent.add(
+                '(pigeonMap[\'${field.name}\'] as num?)?.toDouble()',
+              );
```
I've just removed it because there is no more need.
Needed when the conversion to map used JSONEncoder.

stuartmorgan commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/AllDatatypesTests.swift:
```diff
@@ -0,0 +1,78 @@
+import XCTest
+@testable import Runner
+
+class AllDatatypesTests: XCTestCase {
```
This code needs to follow [the Google Swift style guide](https://google.github.io/swift/); notably, all the indentation here is using 4 spaces instead of two. (We don't have an autoformatter for Swift in our tooling because it's still non-trivial to set up swift-format).

ailtonvivaz replied:
Fixed

stuartmorgan commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/AsyncHandlersTest.swift:
```diff
@@ -0,0 +1,74 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
```
Same for these other files.

ailtonvivaz replied:
FIxed

stuartmorgan commented on packages/pigeon/run_tests.sh:
```diff
@@ -130,6 +130,20 @@ gen_ios_unittests_code() {
     --objc_source_out platform_tests/ios_unit_tests/ios/Runner/$name.gen.m
 }
 
+snake_to_pascal_case() {
```
New tests should be added to run_tests.dart, not run_tests.sh, as the intent is to migrate to run_tests.dart.

ailtonvivaz replied:
Removed

stuartmorgan reviewed (commented):
Looks good to me other than a few high-level notes.

As a caveat, I didn't review the generated output, or look in depth at the generator, since @gaaclarke already did that. But the generator structure, test setup, etc. looks good, and we can iterate on the exact output as we try it out (as has happened with other languages).

stuartmorgan commented:
@ailtonvivaz Are you planning on updating this per the above? It should be very close to landing.

(Once this is in we can start on reviewing the Kotlin generator!)

ailtonvivaz commented:
> @ailtonvivaz Are you planning on updating this per the above? It should be very close to landing.
> 
> (Once this is in we can start on reviewing the Kotlin generator!)

I'll be working on that today.
I hope I can solve all notes.

stuartmorgan commented on packages/pigeon/CHANGELOG.md:
```diff
@@ -30,6 +34,7 @@
 
 * [objc] Simplified some code output, including avoiding Xcode warnings about
   using `NSNumber*` directly as boolean value.
+  
```
Nit: Please revert this stray change.

ailtonvivaz replied:
Revertes

stuartmorgan commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/EchoBinaryMessenger.swift:
```diff
@@ -0,0 +1,46 @@
+  }
+  
+  func send(onChannel channel: String, message: Data?) {
+    
```
These two empty methods could use comments explaining why they are empty.

ailtonvivaz replied:
Added

hellohuanlin commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/NullableReturnsTests.swift:
```diff
@@ -0,0 +1,53 @@
+  var didCall: Bool = false
+  var x: Int32?
+  
+  func doit(x: Int32?) -> Int32 {
```
`doIt`

ailtonvivaz replied:
This was generated for all languages. Do I have change it and the tests for Java, ObjC and C++?

hellohuanlin replied:
i don't have much context around where this comes from. I just review the swift part and noticed this.

stuartmorgan replied:
It's a poor name in all languages, but it's a pre-existing problem and only in test code so let's not worry about it here.

hellohuanlin commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/PrimitiveTests.swift:
```diff
@@ -0,0 +1,230 @@
+    let channelName = "dev.flutter.pigeon.PrimitiveHostApi.anInt"
+    XCTAssertNotNil(binaryMessenger.handlers[channelName])
+    
+    let input: Int = 1
```
`: Int` not necessary

ailtonvivaz replied:
Fixed

hellohuanlin commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/PrimitiveTests.swift:
```diff
@@ -0,0 +1,230 @@
+    let channelName = "dev.flutter.pigeon.PrimitiveHostApi.aBool"
+    XCTAssertNotNil(binaryMessenger.handlers[channelName])
+    
+    let input: Bool = true
```
same here

hellohuanlin commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/Utils.swift:
```diff
@@ -0,0 +1,59 @@
+  
+  guard let x = x as? AnyHashable,
+        let y = y as? AnyHashable else {
+          return false
```
indentation

ailtonvivaz replied:
Fixed

hellohuanlin commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/Utils.swift:
```diff
@@ -0,0 +1,59 @@
+    if equals(x[i], y[i]) == false {
+      return false
+    }
+  }
```
use `allSatisfy`

ailtonvivaz replied:
Would it look like this?
```swift
return (0..<x.count).allSatisfy { equals(x[$0], y[$0]) }
```

hellohuanlin replied:
yeah something like that

ailtonvivaz replied:
Done

hellohuanlin commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/Utils.swift:
```diff
@@ -0,0 +1,59 @@
+    } else {
+      return false
+    }
+  }
```
`allSatisfy`

ailtonvivaz replied:
Would it look like this?
```swift
return x.allSatisfy { equals($0.value, y[$0.key]) }
```
This emit a warning with the message:
**Expression implicitly coerced from 'Any??' to 'Any?'**

hellohuanlin replied:
This means it's wrapped by double optional. You can use `as Any?`.  

That being said, this shouldn't happen in the first place. There is no benefit of having a dictionary of type `[AnyHashable: Any?]` instead of just `[AnyHashable: Any]`

ailtonvivaz replied:
It's needed because we have this situations:
```dart
List<List<bool?>?>? nestedList;
Map<String?, String?>? mapWithAnnotations;
Map<String?, Object?>? mapWithObject;
```

stuartmorgan replied:
> can you elaborate more where do they come from? Why do we use optional as the map value?

Pigeon APIs are defined by the clients of Pigeon; there's no "we" in this context. (Currently Pigeon only supports optional generic types, for somewhat complicated reasons, but even once that's eventually addressed people could still define their APIs this way if they wanted to.)

ailtonvivaz replied:
Done

hellohuanlin reviewed (commented):
Scanned through the Swift part. My main concern is whether we have swift linter up and running to enforce the style guide.

stuartmorgan commented:
We don't have any auto-formatting for Swift at the moment. I looked at it a while ago and it was quite fiddly to get working. IIRC you had to build from source with the exact branch corresponding to the specific Xcode version you were using.

We should look and see if it's gotten better, but automated enforcement of style isn't a blocker for using Swift.

ailtonvivaz commented on packages/pigeon/platform_tests/ios_swift_unit_tests/pubspec.yaml:
```diff
@@ -0,0 +1,68 @@
+# Read more about Android versioning at https://developer.android.com/studio/publish/versioning
+# In iOS, build-name is used as CFBundleShortVersionString while build-number used as CFBundleVersion.
+# Read more about iOS versioning at
+# https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html
```
Removed

ailtonvivaz commented on packages/pigeon/platform_tests/ios_unit_tests/ios/Runner/Info.plist:
```diff
@@ -41,5 +41,7 @@
 	<key>UIViewControllerBasedStatusBarAppearance</key>
 	<false/>
+	<key>CADisableMinimumFrameDurationOnPhone</key>
+	<true/>
```
Removed

ailtonvivaz commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/EchoBinaryMessenger.swift:
```diff
@@ -0,0 +1,46 @@
+    
+  }
+  
+  func send(onChannel channel: String, message: Data?, binaryReply callback: FlutterBinaryReply? = nil) {
```
Fixed

ailtonvivaz commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/EchoBinaryMessenger.swift:
```diff
@@ -0,0 +1,46 @@
+    callback?(codec.encode(firstArg))
+  }
+  
+  func setMessageHandlerOnChannel(_ channel: String, binaryMessageHandler handler: FlutterBinaryMessageHandler? = nil) -> FlutterBinaryMessengerConnection {
```
Fixed

ailtonvivaz commented:
> LGTM with a few nits. Thanks for all the work here!
> 
> This is failing Windows tests in CI, which probably means the Windows tests were broken by changes in `pigeons/`. The CI is reporting it as a hang (we're still looking into why), but locally it'll actually be a compile failure. Do you have a Windows machine to investigate, or do you need one of us to take a look?

@stuartmorgan, I've just fixed the windows tests. There was the need to change some generated class names to adapt the new values.

ailtonvivaz commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/EchoBinaryMessenger.swift:
```diff
@@ -0,0 +1,46 @@
+  func send(onChannel channel: String, message: Data?, binaryReply callback: FlutterBinaryReply? = nil) {
+    guard let args = codec.decode(message) as? [Any?],
+          args.count > 0,
+          let firstArg = args[0],
```
Fixed

ailtonvivaz commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/AsyncHandlersTest.swift:
```diff
@@ -0,0 +1,74 @@
+    binaryMessenger.handlers[channelName]?(nil) { data in
+      let outputMap = binaryMessenger.codec.decode(data) as? [String: Any]
+      XCTAssert(outputMap?["result"] == nil)
+      XCTAssert(outputMap?["error"] == nil)
```
Fixed

ailtonvivaz commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/EchoBinaryMessenger.swift:
```diff
@@ -0,0 +1,46 @@
+    guard let args = codec.decode(message) as? [Any?],
+          args.count > 0,
+          let firstArg = args[0],
+          firstArg is NSNull == false else {
```
Fixed

ailtonvivaz commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/EchoBinaryMessenger.swift:
```diff
@@ -0,0 +1,46 @@
+              callback?(codec.encode(value))
+            } else {
+              callback?(nil)
+            }
```
Fixed

ailtonvivaz commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/HandlerBinaryMessenger.swift:
```diff
@@ -0,0 +1,39 @@
+      return
+    }
+    
+    let result = self.handler(args)
```
Fixed. I've added self in all places

ailtonvivaz commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/HandlerBinaryMessenger.swift:
```diff
@@ -0,0 +1,39 @@
+  
+  func setMessageHandlerOnChannel(_ channel: String, binaryMessageHandler handler: FlutterBinaryMessageHandler? = nil) -> FlutterBinaryMessengerConnection {
+    self.count += 1
+    return .init(self.count)
```
Fixed

hellohuanlin commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/Runner/Assets.xcassets/AppIcon.appiconset/Contents.json:
```diff
@@ -0,0 +1,104 @@
+{
```
Should we remove icons for unit test?

ailtonvivaz replied:
Removed

hellohuanlin commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/PrimitiveTests.swift:
```diff
@@ -0,0 +1,230 @@
+  func aBool(value: Bool) -> Bool { value }
+  func aString(value: String) -> String { value  }
+  func aDouble(value: Double) -> Double { value }
+  func aMap(value: [AnyHashable : Any?]) -> [AnyHashable : Any?] { value }
```
`[Key: Value]` (no white space after the key) per style guide.

ailtonvivaz replied:
Fixed

hellohuanlin reviewed (changes requested):
Request changes for the optional dictionary value type (double optional problem with **Expression implicitly coerced from 'Any??' to 'Any?'**)

hellohuanlin commented:
>  automated enforcement of style isn't a blocker for using Swift.

Gotcha. In this case I think I will do another pass for the swift code.

ailtonvivaz commented on packages/pigeon/lib/pigeon_lib.dart:
```diff
@@ -491,6 +506,28 @@ class JavaGenerator implements Generator {
+  IOSink? shouldGenerate(PigeonOptions options) => _openSink(options.swiftOut);
+
+  @override
+  List<Error> validate(PigeonOptions options, Root root) => <Error>[];
```
I've just implemented this for Swift

stuartmorgan reviewed (approved):
LGTM!

@hellohuanlin Unless you find structural issues, since it's still marked experimental and we can make changes as needed, we can err on the side of landing it. I.e., don't worry about making sure that all generated code for every case is 100% perfect. Traditionally we've found a lot of issues in actual use that weren't evident in review, and then fixed them incrementally as we go.

(Not to say you shouldn't look for issues or that we shouldn't fix them before landing, just that given that the overall structure and test scaffolding are in good shape here, the stakes of landing this PR are relatively low based on our experience with incremental fixes.)

hellohuanlin commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/HandlerBinaryMessenger.swift:
```diff
@@ -0,0 +1,43 @@
+    callback?(self.codec.encode(result))
+  }
+  
+  func setMessageHandlerOnChannel(_ channel: String, binaryMessageHandler handler: FlutterBinaryMessageHandler? = nil) -> FlutterBinaryMessengerConnection {
```
linebreak

ailtonvivaz replied:
FIxed

hellohuanlin commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/HandlerBinaryMessenger.swift:
```diff
@@ -0,0 +1,43 @@
+    }
+    
+    let result = self.handler(args)
+    callback?(self.codec.encode(result))
```
i would do a null check for callback as the first step and skip the whole function if its nil

ailtonvivaz replied:
Done

hellohuanlin commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/EnumTests.swift:
```diff
@@ -0,0 +1,57 @@
+      let outputMap = binaryMessenger.codec.decode(data) as? [String: Any]
+      XCTAssertNotNil(outputMap)
+      
+      let output = outputMap!["result"] as? DataWithEnum
```
avoid force unwrap as much as possible (here and other places)

if this test case fails, we want to continue to the next test case, and not crash the whole test.

ailtonvivaz replied:
But in line 34 the test assert the nullability of variable.
Even with this is it better avoid force unwrap?

hellohuanlin replied:
So how `XCTAssert` works is that, if the assert fails, it will record the failure, and continue to the next line. After running all the tests, you get a report of all the failures in all test cases. However, force unwrap a nil value will simply crash the whole test suite. 

For this particular example, when Line 34 fails, it will go to Line 36, and crash the whole test.

ailtonvivaz replied:
Fixed

hellohuanlin commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/RunnerTests/MockBinaryMessenger.swift:
```diff
@@ -0,0 +1,32 @@
+    }
+  }
+  
+  func setMessageHandlerOnChannel(_ channel: String, binaryMessageHandler handler: FlutterBinaryMessageHandler? = nil) -> FlutterBinaryMessengerConnection {
```
break

ailtonvivaz replied:
Fixed

hellohuanlin reviewed (commented):
Can you double check if the icons and launch screen images are removed? git histories are forever even if you delete these images later.

stuartmorgan commented on packages/pigeon/platform_tests/ios_swift_unit_tests/ios/Runner/Assets.xcassets/LaunchImage.imageset/Contents.json:
```diff
@@ -0,0 +1,21 @@
+  "images" : [
+    {
+      "idiom" : "universal",
+      "filename" : "LaunchImage.png",
```
We have 11 copies of this image in the repository already; there's no meaningful incremental cost to having more copies. We generally don't worry about slimming down the example apps from the template default anyway; having them match the template is the easiest way to maintain them.

stuartmorgan commented:
@ailtonvivaz Once you've had a chance to respond to the final small comments above, let me know it's ready and I can get this landed 🙂

ailtonvivaz commented:
> 

I'm sorry for this mistake. I thought how the PR was approved , all things were "right".
Now, it's fixed.

stuartmorgan commented:
Thanks again for the great contribution!

bladeofgod commented:
hi, i use this for generated swift code, but all class is private. how to make it public ?  ;)

hellohuanlin commented:
@bladeofgod can you open a github issue with your code? PR comment section isn't a great place for discussion.

stuartmorgan commented:
> but all class is private

Please also be sure to include more details when filing the issue about exactly what you are trying to do and what the problem is, because this is definitely not accurate.
